### PR TITLE
Updated Language Evolution for chrono descending, plus clean up

### PR DESCRIPTION
**Pull request by atsansone**
This PR updates the Language Evolution page. I made some language updates, but re-sorted the versions as reverse chronological and added release dates and links to the release announcements.

### CODE REVIEW

**Comment by atsansone** [comment]
Staged at https://dart--staging.web.app/guides/language/evolution

File: src/_guides/language/evolution.md
Comment (parlough) [inline src/_guides/language/evolution.md:305]: I think losing the "always supported" makes this a bit more confusing, since Dart 2.2 has already been out for a long time. If this was a release note, this would make more sense. Perhaps just focus on what was added?

```suggestion
Dart 2.2 added support for **[set literals][]**:
```

File: src/_guides/language/evolution.md
Comment (parlough) [inline src/_guides/language/evolution.md:318]: This text was missing some semantic breaks:

```suggestion
Dart 2.1 added support for **int-to-double conversion**, allowing
developers to set `double` values using integer literals. 
This feature removed the annoyance of 
being forced to use a `double` literal (for example, `4.0`)
when the value was conceptually an integer.
```

File: src/_guides/language/evolution.md
Comment (parlough) [inline src/_guides/language/evolution.md:335]: Cleaning up the semantic breaks here:

```suggestion
Dart 2.0 implemented a new **[sound type system][]**. 
Before Dart 2.0, types weren't fully sound, and 
Dart relied heavily on runtime type checking. 
Dart 1.x code had to be [migrated to Dart 2][].
```

**Review by parlough — CHANGES_REQUESTED** [review]
I've gone through and left some issues, suggestions, and nits. Please feel free to push back or discuss anything.

File: src/_guides/language/evolution.md
Comment (atsansone) [reply src/_guides/language/evolution.md:200]: **comment:** Fixed.

File: src/_guides/language/evolution.md
Comment (atsansone) [reply src/_guides/language/evolution.md:348]: **comment:** I think I have a good update. Have a look.

**Review by atsansone — COMMENTED** [review]
@parlough : Merging my changes after this review.

File: src/_guides/language/evolution.md
Comment (parlough) [inline src/_guides/language/evolution.md:347]: Let's be specific about what kind of version we're talking about. Language versions (which this article is focused on) don't allow you to choose. It's the SDKs themselves (and their included tooling) that do.

```suggestion
In the case of null safety, Dart SDKs 2.12 through 2.19 allowed you
```

File: src/_guides/language/evolution.md
Comment (parlough) [inline src/_guides/language/evolution.md:353]: question: Is it fine to say "check out" twice like this. Would it be better to change the second version to "visit"?

File: src/_guides/language/evolution.md
Comment (parlough) [inline src/_guides/language/evolution.md:357]: Especially as we transition to using caret syntax in pubspec files in the feature, it might make more sense to say something like "the lower bound of the SDK constraint" since the upper and lower bounds won't be specified separately.

File: src/_guides/language/evolution.md
Comment (parlough) [reply src/_guides/language/evolution.md:392]: Reading the previous text, it seems the point of the previous text was that 2.18.3 it doesn't patch the 2.18 language version, it just still is 2.18 in terms of language version. So I'm not sure "patches" is the best choice either.

The feature specification also mentions "A language version is a major/minor version number pair" so it seems there is no patch language versions.

File: src/_guides/language/evolution.md
Comment (parlough) [inline src/_guides/language/evolution.md:430]: Isn't this whole section about how language versioning works?

We could either drop this or be more specific about why they would want to visit the specification.

File: site-shared
Comment (parlough) [inline site-shared:1]: This reverts to an older version of site-shared, please undo this change.

There's a few ways to do this. Two of these easiest might be:

- `git submodule foreach git pull origin main` - Updates each submodule to track their latest commit on `main`. If you named `dart-lang/site-www` something besides `origin`, replace it with that.
- Switch to a branch tracking the latest `origin/main` and run `git submodule update` there.

**Review by parlough — CHANGES_REQUESTED** [review]
Thanks for making some adjustments and responding to my comments and questions!

I'm also happy to discuss synchronously any of the concerns I raised if that works better for you as well.

File: src/_guides/language/evolution.md
Comment (parlough) [reply src/_guides/language/evolution.md]: See [my comments on another issue](https://github.com/dart-lang/site-www/pull/4684#discussion_r1150007415). This page isn't for general release notes and we shouldn't pretend it is in just some situations.

If we want release notes outside the changelog, that likely should be separat.e

File: src/_guides/language/evolution.md
Comment (parlough) [inline src/_guides/language/evolution.md:156]: Sorry I missed this one before I submitted my review, but the text preceding this is not about the operator. Perhaps move this to right after the discussion of the operator.

File: src/_guides/language/evolution.md
Comment (parlough) [inline src/_guides/language/evolution.md:388]: It seems the changes lost the connection to [my comment](https://github.com/dart-lang/site-www/pull/4684#discussion_r1142087005) about this section, but my thoughts still stand:

> I think in general, I prefer the old text here. I think this change makes it say that the "language versions function in a similar manner when they have full compatibility", but the old text is saying that "many language versions are entirely compatible with each other".

File: src/_guides/language/evolution.md
Comment (parlough) [inline src/_guides/language/evolution.md:386]: It seems the changes lost the connection to [my comment](https://github.com/dart-lang/site-www/pull/4684#discussion_r1142087005) about this section, but my thoughts still stand:

> I think in general, I prefer the old text here. I think this change makes it say that the "language versions function in a similar manner when they have full compatibility", but the old text is saying that "many language versions are entirely compatible with each other".

File: src/_guides/language/evolution.md
Comment (atsansone) [reply src/_guides/language/evolution.md]: **comment:** Updated.

File: src/_guides/language/evolution.md
Comment (parlough) [reply src/_guides/language/evolution.md:117]: I do think naturally embedding them may be better overall, but this new header is good to. We can try this out, thanks!

File: src/_guides/language/evolution.md
Comment (parlough) [inline src/_guides/language/evolution.md:147]: These should have an "s" at the end, since its about multiple operators.

Comment (parlough) [inline src/_guides/language/evolution.md:147]: These should have an "s" at the end, since its about multiple operators.

File: src/_guides/language/evolution.md
Comment (parlough) [reply src/_guides/language/evolution.md:192]: Tools are part of the SDK and similar to library changes, they are not relevant when evaluating and understanding language versions.

To be consistent with this focused direction, please remove these tool mentions.

File: src/_guides/language/evolution.md
Comment (parlough) [reply src/_guides/language/evolution.md:198]: Thanks for adding it back! You can always check the [SDK changelog](https://github.com/dart-lang/sdk/blob/main/CHANGELOG.md) or the [archive](https://dart.dev/get-dart/archive) for releases which don't have a release blog.

File: src/_guides/language/evolution.md
Comment (parlough) [inline src/_guides/language/evolution.md:194]: Since you added release dates for all the other entries, consider adding one here.

The [archive](https://dart.dev/get-dart/archive) specifies the binaries were released on Aug 5, 2020.

```suggestion
### Dart 2.9
_Released 5 Aug 2020_
```

File: src/_guides/language/evolution.md
Comment (parlough) [inline src/_guides/language/evolution.md:208]: Remove this tool mention as well to be consistent with the focused direction I mentioned before.

We don't mention tool changes for the other releases.


```suggestion
```

File: src/_guides/language/evolution.md
Comment (parlough) [inline src/_guides/language/evolution.md:240]: I believe it's not that inferences changes when using `Null` now, it's that inference overall changed in this release.

```suggestion
Inference changed when using `Null` values in a `FutureOr` context.
```

File: src/_guides/language/evolution.md
Comment (parlough) [inline src/_guides/language/evolution.md:241]: The `<:` should be outside of the code.

```suggestion
Constraints of forms similar to `Null` <: `FutureOr` now yield `Null`
```

Also we might want to avoid the symbol completely. It's something readers may not be familiar with. Perhaps something like?

```suggestion
Constraints of the form where `Null` is a subtype of `FutureOr` now yield `Null`
```

File: src/_guides/language/evolution.md
Comment (parlough) [inline src/_guides/language/evolution.md:280]: Something doesn't exist as an identifier until you use it as such, so the following may read better:

```suggestion
You can now use `async` as an identifier in asynchronous and generator functions.
```

File: src/_guides/language/evolution.md
Comment (parlough) [inline src/_guides/language/evolution.md:395]: Consider moving the emphasis to "lower bound".

```suggestion
equal to the **lower bound** of the SDK constraint in the `pubspec.yaml` file.
```

File: src/_guides/language/evolution.md
Comment (parlough) [inline src/_guides/language/evolution.md:385]: This general area of text has got to a great point. Thanks for all of the back and forth with me here!

File: src/_guides/language/evolution.md
Comment (parlough) [inline src/_guides/language/evolution.md:409]: Consider dropping this sentence, it doesn't add anything to the meaning and it just seems a bit scary.

```suggestion
```

**Review by parlough — COMMENTED** [review]
It's getting super close. Thanks for all of those changes.

A few last comments and answers to questions:

File: src/_guides/language/evolution.md
Comment (atsansone) [reply src/_guides/language/evolution.md:409]: **comment:** I'll drop it, but I do think it makes any confusion unambiguous.

File: src/_guides/language/evolution.md
Comment (parlough) [inline src/_guides/language/evolution.md:237]: I think the "Constraints of the forms similar to" portion is pretty confusing, can we simplify it? I also didn't notice the missing generic on `FutureOr<T>`. That's important for knowing what the following `T` is referring to.

```suggestion
Constraints where `Null` serves as a subtype of `FutureOr<T>`
now yield `Null` as the solution for `T`.
```

File: src/_guides/language/evolution.md
Comment (parlough) [inline src/_guides/language/evolution.md:394]: This indicates it uses the 2.18 language version by default, not later. It can't use later language versions or else that would break the constraint.

```suggestion
indicates that this package uses the Dart 2.18 language version by default.
```

**Review by parlough — APPROVED** [review]
I had to rebase these changes to fix some issues with the PR and pushed some minor changes. There's **two last changes** that need to be addressed before merging but I want you to take a look. 

Otherwise, looks good to me. Thanks for doing this and all of the discussion. I think the page has got to a really great spot and has a clearer vision now.

Optionally get a second review from someone after making those fixes, but since it's mostly reorganized and rewritten existing content, it should be fine.

**Pull request merged**
